hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java

```diff
@@ -300,11 +304,140 @@ public static String getRemoteUserName() {
     return remoteUser != null ? remoteUser.getUserName() : null;
   }
 
-  public static String getDefaultRatisDirectory(ConfigurationSource conf) {
+  /**
+   * Get the default Ratis directory for a component when the specific
+   * configuration is not set. This creates a component-specific subdirectory
+   * under ozone.metadata.dirs to avoid conflicts when multiple components
+   * are colocated on the same host.
+   *
+   * <p>For backward compatibility during upgrades, this method checks for
+   * existing Ratis data in old locations before using the new component-specific
+   * location. See {@link #findExistingRatisDirectory} for details on old locations.
+   *
+   * @param conf Configuration source
+   * @param nodeType Type of the node component
+   * @return Path to the component-specific ratis directory
+   */
+  public static String getDefaultRatisDirectory(ConfigurationSource conf,
+      HddsProtos.NodeType nodeType) {
     LOG.warn("Storage directory for Ratis is not configured. It is a good " +
             "idea to map this to an SSD disk. Falling back to {}",
         HddsConfigKeys.OZONE_METADATA_DIRS);
     File metaDirPath = ServerUtils.getOzoneMetaDirPath(conf);
-    return (new File(metaDirPath, "ratis")).getPath();
+    
+    // Check for existing Ratis data from old versions for backward compatibility
+    String existingDir = findExistingRatisDirectory(metaDirPath, nodeType);
```

Comment:
@ChenSammi, if `existingDir` is absolutely empty, will it be safe to delete it as well?